### ISSUE TITLE
Problem with table insert function

### ISSUE THREAD

**Issue by stephensong** [robconery/massive-js#60]
Encountered some problems with massive's  table.insert function.  

It works fine if I call it thusly:

    tbl.insert(solnRows); 

(solnRows contains 225 rows, all 225 are inserted ok.) But if I call it like this:

    solnRows.forEach(function (r) { tbl.insert(r); });

it only ever inserts exactly 10 rows, whilst remaining disturbingly silent about what might have happened with the other 215.

Trying to get to the bottom of the issue, I constructed my own  bastardized version of the massive
insert function:

    let insert_rows = function(tblName, data, next) {
        if(!data) throw "insert should be called with data";
        if (!_.isArray(data)) { data = [data]; }

        var delimitedColumnNames = _.map(_.keys(data[0]), function(key){return util.format('"%s"', key);});
        var sql = util.format("INSERT INTO %s (%s) VALUES\n", tblName, delimitedColumnNames.join(", "));
        var rows = _.map(data, function(r) { return "(" + _.map(_.values(r),formatValue).join(", ") + ")"; });
        sql += rows.join(",\n");
        sql += " RETURNING *";
        //console.log(sql);
        pg.connect(url, function (err,client, done) {
            client.query(sql, null, function (err, res) {
                done();  // return client to pool
                if (next) next(err,res);
            });
        });
    };

where formatValue basically calls pg's prepareValue, and then quotes everything but numbers, booleans and nulls.

For some reason, the above function can be called either way, and always inserts the full 225 rows.
So, as yet I've still not been able to come up with any explanation at to why the native massive insert should fail (silently !!! eeek....) the way it does.   Anybody got a clue?

Thanks,

gary

**Comment by robconery** [comment]
First: which version are you using? It looks like 2.0, but I'm not entirely sure.

Second: how can you tell it's "failing silently"? From the code you've offered, you're not examining the err callback (the first argument). I assume you're doing this, but all I have to go on is your example here.

Finally: It looks like you're trying to run an async routine in a loop, which is synchronous. Would you mind showing me the exact routine that's failing? As you've been able to tell - the insert *is working*, but not as you're expecting, which leads me to think it's an async issue.

**Comment by stephensong** [comment]
Hi Rob,

Many thanks for responding, and for massive too.  I like it a lot.  

Yes, I am using Version 2.0

The real code contained a callback  - which was called exactly 10 times, never with an error.

I also suspect it to be 'an async issue', or more likely, me being a dufus.   But when I saw the above code  suffered no such issue I was at a loss to figure out the difference.  Fwiw I also went looking in the postgres logs but found nothing helpful there either.   I know I should really spend more time trying to figure it out, rather than raising half-assed issues, but I am under some time pressure at the moment.  Sorry.

**Comment by stephensong** [comment]
I will revisit later today, and see if I can reproduce it with a standalone example.  Thanks.

**Comment by xivSolutions** [comment]
I can't repro that behavior so far. Running that within `.forEach()` is an odd choice, given that you can pass the array directly to the `insert` method (or even better, the `.save()` method...). However, whether or not it is a "good" thing to do or not, when I run the following against the test database I get all 225 records inserted (although the order in which the inserts are written to the console is . . . odd):

```javascript
function getTestData() { 
  var testArtists = [];
  for(var i = 0; i< 225; i++) { 
    testArtists.push({name: "Artist " + i});
  }
  return testArtists;
};

Massive.connect({connectionString: constr}, function ( err, res) { 
  var newArtists = getTestData();

  newArtists.forEach(function (r) { 
    res.myschema.artists.insert(r, function (err, saved) { 
      console.log(saved);
      // outputs each new record as it is created, with newly inserted serial id
    });
  });
});
```

Maybe I'm missing something here?

**Comment by stephensong** [comment]
Ok,  Thank you both for your attention.  I am grateful.

I've given this all another red hot go, and am still none the wiser.  I've been over and over it, and cannot see what I might be doing wrong.  The only thing I can think of, is that I am running on Windows.   Or might it be that I am running Node with the --harmony flag (I am using Koa for my http server).

Fwiw. here is pretty much all the relevant code:

    "use strict";

    let _ = require("underscore")._;
    let url = process.env.DATABASE_URL;
    let massive = require('massive');
    let pg = require('pg');

    // following copied from pg, as I could not see a way of accessing them directly

    function arrayString(val) {
      var result = '{';
      for (var i = 0 ; i < val.length; i++) {
        if(i > 0) {
          result = result + ',';
        }
        if(val[i] === null || typeof val[i] === 'undefined') {
          result = result + 'NULL';
        }
        else if(Array.isArray(val[i])) {
          result = result + arrayString(val[i]);
        }
        else
        {
          result = result + JSON.stringify(prepareValue(val[i]));
        }
      }
      result = result + '}';
      return result;
    }

    var prepareValue = function(val, seen) {
      if (val instanceof Buffer) {
        return val;
      }
      if(val instanceof Date) {
        return dateToString(val);
      }
      if(Array.isArray(val)) {
        return arrayString(val);
      }
      if(val === null || typeof val === 'undefined') {
        return null;
      }
      if(typeof val === 'object') {
        return prepareObject(val, seen);
      }
      return val.toString();
    };

    function prepareObject(val, seen) {
      if(val.toPostgres && typeof val.toPostgres === 'function') {
        seen = seen || [];
        if (seen.indexOf(val) !== -1) {
          throw new Error('circular reference detected while preparing "' + val + '" for query');
        }
        seen.push(val);

        return prepareValue(val.toPostgres(prepareValue), seen);
      }
      return JSON.stringify(val);
    }

    function dateToString(date) {
      function pad(number, digits) {
        number = ""+number;
        while(number.length < digits)
          number = "0"+number;
        return number;
      }

      var offset = -date.getTimezoneOffset();
      var ret = pad(date.getFullYear(), 4) + '-' +
        pad(date.getMonth() + 1, 2) + '-' +
        pad(date.getDate(), 2) + 'T' +
        pad(date.getHours(), 2) + ':' +
        pad(date.getMinutes(), 2) + ':' +
        pad(date.getSeconds(), 2) + '.' +
        pad(date.getMilliseconds(), 3);

      if(offset < 0) {
        ret += "-";
        offset *= -1;
      }
      else
        ret += "+";

      return ret + pad(Math.floor(offset/60), 2) + ":" + pad(offset%60, 2);
    }

    // my own little helper

    let formatValue = function(x) {
        if(x === null || typeof x === 'undefined') return 'null';
      
        let y = prepareValue(x);
        let typ = typeof x;

        if (typ != 'number' && typ != 'boolean') y = "'" + y + "'";
        return y
    };

    let insert_rows = function(tblName, data, next) {
        if(!data) throw "insert should be called with data";
        if (!_.isArray(data)) { data = [data]; }

        var delimitedColumnNames = _.map(_.keys(data[0]), function(key){return util.format('"%s"', key);});
        var sql = util.format("INSERT INTO %s (%s) VALUES\n", tblName, delimitedColumnNames.join(", "));
        var rows = _.map(data, function(r) { return util.format("(%s)", _.map(_.values(r),formatValue).join(", ") ); });
        sql += rows.join(",\n") + "\n RETURNING *";
        //console.log(sql);
        pg.connect(url, function (err,client, done) {
            client.query(sql, null, function (err, res) {
                done();  // return client to pool
                if (!err) rslt = res.rows[0];
                if (next) next(err,rslt);
            });
        });
    };

    //  I added the following loadSync routine in manually, as it wasn't in the npm release.
    //  with the following additional line at the top:
    //   'var result = null;'
    //  as it otherwise referred to a global result var.
    //  I tried replacing this with the connection syntax used in xivSolutions' example above, but got
    //  precisely the same results :-(
    let db = massive.loadSync({connectionString: url});

    let insert_solution = function (soln) {
        // this line executes correctly (225 times)
        console.log("inserting solution for : " + soln.puzzle);  

        // this line works fine - all 225 rows inserted
        insert_rows('solutions', soln, function(err, rslt) {    

        // if we comment out the above line, and replace it with the one below,
        // we only ever insert 10 records, and the callback is only called with the successes.
        //db.solutions.insert( soln, function(err, rslt) {  

            if (err) console.log("failed : " + err);  // never executes, despite 215 records going missing 
            else console.log("solution inserted for : " + soln.puzzle);
        });
    };

    let solns = // an array of, in my case, 225 rows loaded from elsewhere

    //  the technique of calling insert just once with all rows batched together works fine 
    // either way, so  I have not included it here.
    solns.forEach(function (r) { insert_solution(r); });

**Comment by xivSolutions** [comment]
The `massive.loadSync()` is an experimental thing (although I see no reason it shouldn't work, given what you've supplied here. On the other hand, bear in mind all it does is load the db synchronously - it does not allow you to use Massive itself synchronously beyond that (and I should document it as such!). 

just for kicks, can you try (if you haven't already) ditching the call to `.loadsync()` and do something like this:

```javascript
Massive.connect({connectionString: constr}, function ( err, res) { 

  // Is there a callback that can happen here when your load is complete? this appears to 
  // possibly be called synchronously, but if it is an async process, might cause issues:
  let solns = // an array of, in my case, 225 rows loaded from elsewhere

  solns.forEach(function (r) {
    db.solutions.insert(r, function (err, saved) { 
      console.log(saved);
      // outputs each new record as it is created, with newly inserted serial id
    });
  });
});
```

From here it still looks like there is an odd mix of sync and async going on here. depending on how you are loading the records into `solns` "from elsewhere", I'm wondering if you aren't cooking right through that, and beginning loop iteration before the load is complete. On the other hand, it does seem your call to `insert_solution()` is actually executing 225 times. 

No matter how you slice it though, calling `insert()` within the `forEach` loop is an odd, less-efficient approach, and also requires fetching a new connection from the pool for each insert. Is there a reason you are taking this approach?

**Comment by robconery** [comment]
The problem is async, I'm 99% sure. You can't run async code in a loop and expect it to work all the time. In some cases (like with a console app that stays open and doesn't exit early) it *might look like it works*, but that's only because the Node process is staying alive until you Ctrl-c it.

The best way to handle this is to build your array of inserts first, then push them in. Preferably in a transaction.

Think about the nature of callbacks and the Event Loop. You're invoking something that does some IO and is expecting a callback to notify you when it's completed. The for loop executes and each one of those steps gets pushed onto the Event Loop - your callback is the only way for your app to know the call was executed and has completed. If the Node process stops, then yeah everything grinds to a halt.

In short: *I am 99.999999% sure this is an async problem* though I can't prove it, unfortunately. You can fix this by passing a built array.

**Comment by stephensong** [comment]
Thanks to you both.  To John: yes I tried what you suggested (as indicated in my comments in the code), and it changed not one iota.  I have tried (and removed) the deasync thing in several contexts now, and I am satisfied that it itself is not the source of the issue.

If it is indeed 'an async issue', then why does the code I posted work where the massive code doesn't, when they _appear_ to be entirely equivalent from the async / sync perspective.  Given that neither of you can reproduce the issue I guess we will have to just leave it there.  The issue is real though - for me at least, and I just hope that an explanation will present itself at some point in the future. Until then, I will be feeling like I am building castles on sand.  Whilst it is true that in this particular case there is an obvious and workable alternative, one can easily envisage circumstances where such was not the case.   Should I close the issue?

**Comment by xivSolutions** [comment]
@stephensong - I can't speak for Rob, but I intend to try this on my Windows machine in the near future (I was working on the MBP when I tested). While I agree with Rob in that it seems like a funny mixture of sync/async (and an edge case), I very much want to see if it can be repro'd. 

I caught that in your comments, but wanted to make sure you were running the precise same code I was, and not using the deasync stuff. 

I think closing the issue for now is probably appropriate, but that doesn't mean investigation stops (for me, anyway). I think there is value to be had in understanding what's going on, even if that is to determine for certain that it is, in fact, a sync issue.

**Comment by robconery** [comment]
@stephensong Can't say why, unfortunately. I do think it's interesting and yes @xivSolutions if you want to dig in by all means do! In general, though, my thinking on this (for now) is 1) executing a DB in an unbound loop is almost always a bad idea and 2) executing async calls in a loop will almost always cause pain.

Looking over your code - is prepping an array an option for you? That would solve your problems and also (most likely) make things a bit more performant. In addition, the batch gets executed within a transaction so if there's a failure you don't have partial inserts.

**Comment by stephensong** [comment]
Ok.  Back from a break, I have looked into this yet again, and have found what is causing my problem.  In the file runner.js, starting around line 30, the following lines appear:

    pg.connect(this.connectionString, function (err, db, done) {
      //throw if there's a connection error
      assert.ok(err === null, err);

      db.query(args.sql, args.params, function (err, result) {
        //we have the results, release the connection
        done(db);

If I change the last line to remove the parameter, i.e.

      done();

my problem disappears!

I have tried this out in my own code (as posted previously), and the (mis)behaviour is entirely consistent.  If the parameter is supplied to the provided 'done' function, then I only ever see 10 records inserted.   If not, everything is hunky-dory. 

Obvious questions remaining:  
- Why is the parameter supplied in the done call?   The example in the pg wiki doesn't supply it.
- Should it be there or not?  
- What is it supposed to achieve? 
- What is the downside of not supplying it? 
and:
- Why am I the only one seeing this problem?

For the moment I am going to remove it in my local copy, and hope nothing bad happens.

Thanks.

**Comment by robconery** [comment]
Wow thanks for the good work here! I think I now have a firm grasp on this :). The reason you only see 10 records is this:

https://github.com/brianc/node-postgres/blob/master/lib/defaults.js#L27

You're using up all the connections and maxing out the pool. Once that happens (I'm guessing) the writes stop because there are no more connections to be had. When you pass the client instance back through `done()` as we're doing, it's supposed to release it back to the pool (meaning we don't need it anymore). Here's an example:

https://github.com/brianc/node-postgres/wiki/Example

I think this might be a bug in `pg`. We could also just use `db.end()` but that means we won't have pooling. Which I think might be OK because we're working in a single-threaded environment and I can't see any way in which we could have two connections open at once.

Originally we had just `done()` in there but it was slowing our tests down (for some reason). @xivSolutions will know more on this - it was something he spotted.

You're the only one seeing this (I think) because executing in a loop as you're doing is not a typical operation... and I think it's an edge case to be honest but then again, that's where the bugs are yes?

@xivSolutions would you mind switching over to `db.end()` and rerunning the tests? I'm on horrible internet here again...

Thanks again @stephensong for your investigation!

**Comment by xivSolutions** [comment]
Ahhh...

The call to `done(db)` has, in fact, been fixed in `master` (see #43 )but not pushed to NPM. When you pass a "truthy" value to `done()` it destroys the client and removes it from the pool, causing both performance issues, and the issue you are seeing here. I assume your code is depleting the available PG clients.

From briancs' wiki: https://github.com/brianc/node-postgres/wiki/pg#parameters

From the [previous issue](https://github.com/robconery/massive-js/issues/43):

According to brianc's wiki, calling done() returns the client to the pool, ready for another connection, but calling `done(db)` (or passing any "truthy" value) actually causes the client to be removed from the pool and and destroyed. 

@robconery - This has been fixed but not pushed to NPM. However, Interesting idea to change it to `end()`. I'll try that, and see what happens. Meantime, though, we have a number of issues fixed that probably warrant a new NPM version as soon as possible, when you're able. 

Specifically #43 (directly related to this one (the change from `done(db)` to `done()`)  and #55 (`tables.sql` returning multiple PK columns)  are fairly important. 

I'll do some messing about with `end()` instead of `done()` and update this thread this evvening. I'm at a work conference in lovely Bend, OR right now ...

**Comment by robconery** [comment]
Pushed just now - thanks John I think I spaced on that :). @stephensong if you pull 2.0.4 you should see your error go away :).
